File: examples/ch08/py/before.py
"""Chapter 8 · BEFORE: closed for extension, open for modification — backwards.

Every new payment provider means editing process_payment — the one function
you least want to touch, and the one whose existing branches you most risk
breaking.
"""

from dataclasses import dataclass


@dataclass(frozen=True)
class Receipt:
    provider: str
    amount: float
    confirmation: str


def process_payment(method: str, amount: float) -> Receipt:
    if method == "card":
        return Receipt("card", amount, f"card-{int(amount * 100)}")
    elif method == "paypal":
        return Receipt("paypal", amount, f"pp-{int(amount * 100)}")
    else:
        raise ValueError(f"unknown payment method: {method}")

File: examples/ch08/ts/before.ts
// Chapter 8 · BEFORE: closed for extension, open for modification — backwards.
//
// Every new payment provider means editing processPayment — the one function
// you least want to touch, and the one whose existing branches you most risk
// breaking.

export interface Receipt {
  readonly provider: string;
  readonly amount: number;
  readonly confirmation: string;
}

export function processPayment(method: string, amount: number): Receipt {
  switch (method) {
    case "card":
      return { provider: "card", amount, confirmation: `card-${Math.round(amount * 100)}` };
    case "paypal":
      return { provider: "paypal", amount, confirmation: `pp-${Math.round(amount * 100)}` };
    default:
      throw new Error(`unknown payment method: ${method}`);
  }
}
